### DISCUSSION TITLE
Possible to eager load in select using hybrid expression?

### DISCUSSION THREAD

**Discussion by dusktreader** [sqlalchemy/sqlalchemy#10120] [Usage Questions] [answered]
I have an ORM class derived from `Base` with a 1-M relationship to another. Futher, I've defined a hybrid attribute (with an analog expression) to compute a sum based on a numerical column in the related class. I am able to get a list of matching ORM objects to load, but without calling `refresh()` on each one, the hybrid attribute won't load. Ideally, I would like to be able to load the hybrid attribute eagerly at the first query execution time.

Example code:

```python
# stub of https://github.com/omnivector-solutions/license-manager/blob/main/backend/lm_backend/api/models/feature.py
class Feature(Base):
    __tablename__ = "features"
    id = Column(Integer, primary_key=True)
    bookings = relationship("Booking", back_populates="feature", lazy="selectin")

    @hybrid_property
    def booked_total(self) -> int:
        return sum(b.quantity for b in self.bookings)

    @booked_total.inplace.expression
    @classmethod
    def _booked_total_expression(cls) -> SQLColumnExpression[int]:
        return select(func.sum(Booking.quantity)).where(Booking.feature_id == cls.id).label("booked_total")
```

```python
# stub of https://github.com/omnivector-solutions/license-manager/blob/main/backend/lm_backend/api/models/booking.py
class Booking(Base):
    __tablename__ = "bookings"
    id = Column(Integer, primary_key=True)
    quantity = Column(Integer, CheckConstraint("quantity>=0"), nullable=False)
```

Now, when I execute a query `without` refreshing the resulting the resulting objects, the `booked_total` attribute shows `0`:

```python
# stub of https://github.com/omnivector-solutions/license-manager/blob/main/backend/lm_backend/api/cruds/generic.py

async def read_all(self, db_session: AsyncSession, force_refresh: bool = False):
    stmt = select(self.model)
    query = await db_session.scalars(stmt)
    db_objs = list(query.all())
    if force_refresh:
        await gather(*(db_session.refresh(db_obj) for db_obj in db_objs))
    return db_objs
```

When I analyzed the query that's produced from the select, it does not join against `bookings`. However, if I include the hybrid attribute as a column in the select, it will do the join. The `scalars()` method doesn't work, in this case, though because executing returns a tuple of the Feature object and a scalar `booking_total`.

The only way I have found to get the `booked_total` to populate is to use the `force_refresh` parameter to the method and refresh each individual result.

In a case like this where I know I want the `booked_total` to be included in the result of a query, I would love to be able to direct sqlalchemy to use the hybrid expression eagerly and to load the results in the resulting ORM objects.

Based on some investigation (see: https://groups.google.com/g/sqlalchemy/c/XQW8gxX93hU), it seems like it might be a limitation of hybrid attributes that a select can't load them into the dict of the resulting objects. Is there a way to work around this? I would really like to reap the performance gains of including the sub-select in the generated query instead of issuing a bunch of axillary queries to load the computed value.

Thanks!!

**Comment by zzzeek** [comment]
hi -

a hybrid attribute, when you call it from a loaded instance, is just a python function, a method on the instance.  I dont see any example use here in the code given, but if you have code that says `assert my_instance.booked_total > 0` you could replace that with `assert sum(b.quantity for b in my_instance.bookings) > 0` and that is equivalent.   

When you say "However, if I include the hybrid attribute as a column in the select, it will do the join." that doesnt match the code illustrated because you have `lazy="selectin"`, which normally should just be SELECTing from "bookings" alone, with a list of primary key identifiers from "features".   if the relationship is a simple one-to-many it does not need to emit a JOIN to the parent table.

overall, I would be able to provide better guidance if the above fragments were worked into a self-contained [MCVE](http://stackoverflow.com/help/mcve) that I can copy-and-paste here and run.

**Reply by zzzeek** [reply, marked as answer]
Here's an MCVE.  See if you can modify this to show the mis-behavior you are seeing

```py
import asyncio

from sqlalchemy import CheckConstraint
from sqlalchemy import Column
from sqlalchemy import ForeignKey
from sqlalchemy import func
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy import SQLColumnExpression
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship

Base = declarative_base()


class Feature(Base):
    __tablename__ = "features"
    id = Column(Integer, primary_key=True)
    bookings = relationship(
        "Booking", back_populates="feature", lazy="selectin"
    )

    @hybrid_property
    def booked_total(self) -> int:
        return sum(b.quantity for b in self.bookings)

    @booked_total.inplace.expression
    @classmethod
    def _booked_total_expression(cls) -> SQLColumnExpression[int]:
        return (
            select(func.sum(Booking.quantity))
            .where(Booking.feature_id == cls.id)
            .label("booked_total")
        )


class Booking(Base):
    __tablename__ = "bookings"
    id = Column(Integer, primary_key=True)
    quantity = Column(Integer, CheckConstraint("quantity>=0"), nullable=False)
    feature_id = Column(ForeignKey("features.id"))
    feature = relationship("Feature", back_populates="bookings")


async def read_all(db_session: AsyncSession):
    stmt = select(Feature)
    query = await db_session.scalars(stmt)
    db_objs = list(query.all())
    return db_objs


async def generate_data(db_session: AsyncSession):
    db_session.add_all(
        [
            Feature(
                bookings=[
                    Booking(quantity=5),
                    Booking(quantity=10),
                ]
            ),
            Feature(
                bookings=[
                    Booking(quantity=8),
                    Booking(quantity=19),
                ]
            ),
        ]
    )
    await db_session.commit()


async def main():
    engine = create_async_engine(
        "postgresql+asyncpg://scott:tiger@localhost/test", echo=True
    )

    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)
        await conn.run_sync(Base.metadata.create_all)

    async with AsyncSession(engine) as sess:
        await generate_data(sess)

    async with AsyncSession(engine) as sess:
        result = await read_all(sess)

        for obj in result:
            print(f"Object: {obj} booked total: {obj.booked_total}")
            assert obj.booked_total > 0


asyncio.run(main())

```


output within the read_all() function:

```
2023-07-19 09:45:29,035 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2023-07-19 09:45:29,036 INFO sqlalchemy.engine.Engine SELECT features.id 
FROM features
2023-07-19 09:45:29,036 INFO sqlalchemy.engine.Engine [generated in 0.00024s] ()
2023-07-19 09:45:29,043 INFO sqlalchemy.engine.Engine SELECT bookings.feature_id AS bookings_feature_id, bookings.id AS bookings_id, bookings.quantity AS bookings_quantity 
FROM bookings 
WHERE bookings.feature_id IN ($1::INTEGER, $2::INTEGER)
2023-07-19 09:45:29,043 INFO sqlalchemy.engine.Engine [generated in 0.00028s] (1, 2)
Object: <__main__.Feature object at 0x7fcb1727a910> booked total: 15
Object: <__main__.Feature object at 0x7fcb1727a990> booked total: 27
2023-07-19 09:45:29,046 INFO sqlalchemy.engine.Engine ROLLBACK

```

**Reply by zzzeek** [reply]
if these objects were already loaded, perhaps you want to add [populate_existing](https://docs.sqlalchemy.org/en/20/orm/queryguide/api.html#orm-queryguide-populate-existing) ?

```py
async def read_all(db_session: AsyncSession):
    stmt = select(Feature).execution_options(populate_existing=True)
    query = await db_session.scalars(stmt)
    db_objs = list(query.all())
    return db_objs
```

**Reply by dusktreader** [reply]
Thank you for your response on this. I'm sorry I got busy and never got back to you with an MCVE. Thank you for taking the time to explain this and provide your own example.

**Reply by mdpdesign** [reply]
Hi @zzzeek I'm having similar problem/question, as I'd like to count related objects in my model. I've read the documentation on hybrid attributes and `column_property` but being honest I can't figure it out for my use case.

In my queries for Status I don't always want to load all related objects (Order), but I always want to return the count of Order(s) as Status attribute. I'm having trouble to configure that behavior with hybrid attributes, but it works using `column_property`, however then this attribute is not recognized as class member by IDE etc.

This is modified code from above which is very similar to my production code:

```python
import asyncio
import urllib.parse

from sqlalchemy import CheckConstraint
from sqlalchemy import Column
from sqlalchemy import ForeignKey
from sqlalchemy import func
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy import SQLColumnExpression
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.orm import column_property
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import noload
from sqlalchemy.orm import relationship


class Base(DeclarativeBase):
    pass


class Status(Base):
    __tablename__ = "statuses"
    id = Column(Integer, primary_key=True)

    orders = relationship("Order", back_populates="status")

    @hybrid_property
    def order_count(self) -> int:
        return len(self.orders)

    # Trying to add expression for hybrid property, but it is not recognized by IDE etc.
    # and also not working when querying
    @order_count.inplace.expression
    @classmethod
    def _order_count_expression(cls) -> SQLColumnExpression[int]:
        stmt = (
            select(func.count(Order.id))
            .where(Order.status_id == cls.id)
            .correlate_except(Order)
            .scalar_subquery()
            .label("order_count")
        )
        return stmt


class Order(Base):
    __tablename__ = "orders"
    id = Column(Integer, primary_key=True)
    quantity = Column(Integer, CheckConstraint("quantity>=0"), nullable=False)
    status_id = Column(ForeignKey("statuses.id"))

    status = relationship("Status", back_populates="orders")

# Adding column property to Status model works and count is returned properly,
# but attribute is not recognized by IDE etc.
Status.order_count_column_property = column_property(
    select(func.count(Order.id)).where(Order.status_id == Status.id).correlate_except(Order).scalar_subquery()
)


async def read_all(db_session: AsyncSession):
    stmt = select(Status).options(
        # Here I don't always want to load all related orders and their full data
        # but I still want to count them as subquery as defined in hybrid property or coulmn_property
        noload(Status.orders),
    )
    query = await db_session.scalars(stmt)
    db_objs = list(query.all())
    return db_objs


async def generate_data(db_session: AsyncSession):
    db_session.add_all(
        [
            Status(
                orders=[
                    Order(quantity=5),
                    Order(quantity=10),
                ]
            ),
            Status(
                orders=[
                    Order(quantity=8),
                    Order(quantity=19),
                ]
            ),
        ]
    )
    await db_session.commit()


async def main():
    db_user: str = "postgres"
    db_password: str = urllib.parse.quote_plus("P@ssw0rd")
    db_host: str = "postgres"
    db_name: str = "test"
    engine = create_async_engine(f"postgresql+asyncpg://{db_user}:{db_password}@{db_host}/{db_name}", echo=True)

    async with engine.begin() as conn:
        await conn.run_sync(Base.metadata.drop_all)
        await conn.run_sync(Base.metadata.create_all)

    async with AsyncSession(engine) as sess:
        await generate_data(sess)

    async with AsyncSession(engine) as sess:
        result = await read_all(sess)

        for obj in result:
            print(
                f"Object: {obj} order count: {obj.order_count}, order order_count_column_property: {obj.order_count_column_property}"
            )


asyncio.run(main())
```

Here is the script output:

```
2025-10-15 12:55:00,082 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2025-10-15 12:55:00,083 INFO sqlalchemy.engine.Engine INSERT INTO statuses (id) VALUES (DEFAULT), (DEFAULT) RETURNING statuses.id, statuses.id AS id__1
2025-10-15 12:55:00,083 INFO sqlalchemy.engine.Engine [generated in 0.00003s (insertmanyvalues) 1/1 (ordered)] ()
2025-10-15 12:55:00,085 INFO sqlalchemy.engine.Engine INSERT INTO orders (quantity, status_id) SELECT p0::INTEGER, p1::INTEGER FROM (VALUES ($1::INTEGER, $2::INTEGER, 0), ($3::INTEGER, $4::INTEGER, 1), ($5::INTEGER, $6::INTEGER, 2), ($7::INTEGER, $8::INTEGER, 3)) AS imp_sen(p0, p1, sen_counter) ORDER BY sen_counter RETURNING orders.id, orders.id AS id__1
2025-10-15 12:55:00,085 INFO sqlalchemy.engine.Engine [generated in 0.00003s (insertmanyvalues) 1/1 (ordered)] (5, 1, 10, 1, 8, 2, 19, 2)
2025-10-15 12:55:00,086 INFO sqlalchemy.engine.Engine COMMIT
2025-10-15 12:55:00,086 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2025-10-15 12:55:00,087 INFO sqlalchemy.engine.Engine SELECT statuses.id, (SELECT count(orders.id) AS count_1 
FROM orders 
WHERE orders.status_id = statuses.id) AS anon_1 
FROM statuses
2025-10-15 12:55:00,087 INFO sqlalchemy.engine.Engine [generated in 0.00012s] ()
Object: <__main__.Status object at 0xffffaa4df110> order count: 0, order order_count_column_property: 2
Object: <__main__.Status object at 0xffffaa4b95b0> order count: 0, order order_count_column_property: 2
2025-10-15 12:55:00,088 INFO sqlalchemy.engine.Engine ROLLBACK
```

This is one of the "production" functions I use to query Status, by default I'm not loading relationships, only when I need it:

```python
async def get_status(
    self,
    status_id: int,
    **kwargs,
) -> Status:
    try:
        stmt = select(Status).where(Status.id == status_id)

        orders_options: tuple = (noload(Status.orders),)
        transitions_options: tuple = (noload(Status.transitions_to),)

        if kwargs.get("with_orders"):
            orders_options = (selectinload(Status.orders),)

        if kwargs.get("with_transitions"):
            transitions_options = (
                selectinload(Status.transitions_to).options(
                    joinedload(
                        StatusTransitionAssociation.to_status,
                        innerjoin=True,
                    ),
                ),
            )

        stmt = stmt.options(*orders_options, *transitions_options)

        status_obj: Status | None = await self._session.scalar(stmt)

        if not status_obj:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Status with ID {status_id} not found",
            )

        return status_obj
    except Exception:
        logger.error("Failed to get status with ID: %s", status_id)
        raise
```

Is it possible to use hybrid attributes as per my use case or do I have to use `column_property` only? Maybe I'm doing something wrong? 

I appreciate any help

**Reply by zzzeek** [reply]
> In my queries for Status I don't always want to load all related objects (Order), but I always want to return the count of Order(s) as Status attribute. I'm having trouble to configure that behavior with hybrid attributes, but it works using `column_property`, however then this attribute is not recognized as class member by IDE etc.

use column property and give it a type annotation, IDE will see it fine

```
from sqlalchemy.orm import Mapped

class Status(Base):
    __tablename__ = "statuses"
    id = Column(Integer, primary_key=True)

    orders = relationship("Order", back_populates="status")

    order_count_column_property: Mapped[int]

```

**Reply by mdpdesign** [reply]
Hi @zzzeek thank you for your reply, I haven't thought to do it like this. 🤦‍♂️ - I will. 

But anyway the thing I wanted to ask actually is whether it is required to use `column_property` in such case when also using `noload` for relationships instead of hybrid attributes? - I'm seeking for best practices on this and perhaps few words of explanation why one over the other.

Also for the `noload` (which is btw. considered **legacy** as per current docs 😬) is it that we should expect deprecation for this load technique?

I'm `noload`'ing some of my relationships for e.g. for performance reasons and because when FastAPI or Pydantic are serializing models into schemas they tend to access attributes that I can no longer fetch from DB because those attributes are accessed outside of **AsyncSession** - And usually I was getting **MissingGreenlet** errors, or similar -  https://docs.sqlalchemy.org/en/20/errors.html#missinggreenlet

I'm wondering what are the best practices for things like that as well? ☺️

**Reply by zzzeek** [reply]
noload is not a thing you should use because it tells a relationship that if it's ever accessed, it should return the wrong result (None).  so I wouldnt use it.  just dont load the relationships you dont want to lazy load, or have them use raiseload so they raise when accessed.    A more robust solution for lazy attributes under asyncio is to use [AsyncAttrs](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncAttrs).

if you want an ORM query for `Status` alone, like `select(Status)`, to also perform a subquery that gets this count value, then yes you need to use column_property() for that.  with the hybrid, to get your expression to run you'd need to say `select(Status, Status.my_hybrid_expression)`, which would include the subquery but it would not be applied to the `Status` object, it would just be another column.

**Reply by mdpdesign** [reply]
> noload is not a thing you should use because it tells a relationship that if it's ever accessed, it should return the wrong result (None). so I wouldnt use it. just dont load the relationships you dont want to lazy load, or have them use raiseload so they raise when accessed. A more robust solution for lazy attributes under asyncio is to use [AsyncAttrs](https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncAttrs).

I totally get your point for the `noload` and I would not use it, but the "problem" with combination of FastAPI, Pydantic & Async SQLAlchemy is that when SQLAlchemy models are turned into Pydantic (V2) schemas is that Pydantic is accessing all model attributes when it creates & validates the object. It is better perhaps described here - https://github.com/pydantic/pydantic/issues/8192 and this comment https://github.com/pydantic/pydantic/issues/8192#issuecomment-2574355699

In general because Pydantic is accessing SQLAlchemy model objects properties (including relationships) when FastAPI is returning the response (outside my code & my control) it causes lazy load outside AsyncSession, thus we get "MissingGreenlet" or "Parent instance <x> is not bound to a Session; (lazy load/deferred load/refresh/etc.) operation cannot proceed" errors.

So I see myself and other people also as per Pydantic issue above to use `noload` trick, so that DB model attribute is set to even wrong result like `None` or empty list `[]` just to not cause any extra lazy load later by Pydantic upon returning the response. I hope you get the point 😄

As per the Pydantic issue, if there would be some option to exclude some attributes when "transforming" DB model to schema, I guess that could solve the issue (because we could exclude relationship attributes). Anyway I don't think this is an issue with SQLAlchemy, I was worried about deprecating `noload` from the library so we will be left w/o the workaround to that problem.

> if you want an ORM query for Status alone, like select(Status), to also perform a subquery that gets this count value, then yes you need to use column_property() for that. with the hybrid, to get your expression to run you'd need to say select(Status, Status.my_hybrid_expression), which would include the subquery but it would not be applied to the Status object, it would just be another column.

I actually thought about that and even tested it with the "test" code I pasted in my previous reply, but when combined with `noload` option to the query, hybrid attribute isn't set 🙃

Result of the query:

```
Object: <__main__.Status object at 0xffffaa4df110> order count: 0, order order_count_column_property: 2
Object: <__main__.Status object at 0xffffaa4b95b0> order count: 0, order order_count_column_property: 2
```

Anyway, thank you for the great work and taking the time to answer these questions ☺️

**Reply by zzzeek** [reply]
> I totally get your point for the `noload` and I would not use it, but the "problem" with combination of FastAPI, Pydantic & Async SQLAlchemy is that when SQLAlchemy models are turned into Pydantic (V2) schemas is that Pydantic is accessing all model attributes when it creates & validates the object. 

im aware of it, i think it's a silly design (at least that it can't be controlled in any way, seemingly) :)

> In general because Pydantic is accessing SQLAlchemy model objects properties (including relationships) when FastAPI is returning the response (outside my code & my control) it causes lazy load outside AsyncSession, thus we get "MissingGreenlet" or "Parent instance is not bound to a Session; (lazy load/deferred load/refresh/etc.) operation cannot proceed" errors.

you know when I look at all the docs for FastAPI I dont see asyncio hardly being used at all.  https://fastapi.tiangolo.com/tutorial/sql-databases/#create-a-hero 

how is everyone that uses FastAPI coming here doing everything in asyncio?    It's not even required AFAICT


> 
> So I see myself and other people also as per Pydantic issue above to use `noload` trick, so that DB model attribute is set to even wrong result like `None` or empty list `[]` just to not cause any extra lazy load later by Pydantic upon returning the response. I hope you get the point 😄

Yeah noload is still there and clearly we can't remove it but sure seems strange to have an attribute on your model that you dont need right ? :)    lazyloading is not SQLAlchemy's invention so it's weird pydantic has seemingly zero approach for model objects that have optional, expensive-to-use attributes on them

> As per the Pydantic issue, if there would be some option to exclude some attributes when "transforming" DB model to schema, I guess that could solve the issue (because we could exclude relationship attributes). 

yeah exactly

**Reply by mdpdesign** [reply]
Yeah it's hard to not agree with you @zzzeek 🤣 about Pydantic not taking that into account

> you know when I look at all the docs for FastAPI I dont see asyncio hardly being used at all. https://fastapi.tiangolo.com/tutorial/sql-databases/#create-a-hero
>
> how is everyone that uses FastAPI coming here doing everything in asyncio? It's not even required AFAICT

Yeah, that is actually interesting, It might be my lack of knowledge, I'm learning things.. or perhaps wrong assumptions that I really should do everything async. Even though FastAPI claims that it can handle both `def func()` and `async def func()` properly.

Do you suggest to switch to regular functions and not use any async functions? I'm wondering about performance then.. I would need to test that out.

In the mean time, I came with such an idea for Pydantic transform problem, instead of passing DB model directly I use `inspect()` to get `dict` for the model and pass that to Pydantic schema class (implicitly when returning from function), `dict` is simpler and does not perform any unwanted attributes access outside AsyncSession 🤷‍♂️:

```python
@status_router.post(
    "/",
    summary="Create a new status",
    status_code=status.HTTP_201_CREATED,
    response_model=StatusOutNotNested,
)
async def create_status(
    repo: StatusRepositoryDep,
    status_data: StatusCreate,
):
    new_db_status: Status = await repo.create_status(status_data)
    insp: InstanceState = inspect(new_db_status)

    # Returning dict to avoid Pydantic from trying to parse
    # SQLAlchemy relationships which are not loaded
    return insp.dict
```

I'm not sure if this is even a good approach, currently I'm trying to test that. 

Alongside I had to add `order_count` to Pydantic schema model with default value set to `0`. This is because DB model `dict` doesn't have this attribute when creating it in DB - related to `column_property` usage. So then I had Pydantic validation error that it was missing.. 🤣

```python
class StatusOutNotNested(StatusOutBase):
    # Other attributes...

    order_count: int = Field(default=0)

    model_config = ConfigDict(from_attributes=True)
``` 

Btw. is there any reason not to use `Status.__dict__` directly from DB model? I used `inspect()` to have more "public" access to it.

Again thank you for all your input and help!

**Reply by CaselIT** [reply]
> Btw. is there any reason not to use `Status.__dict__` directly from DB model? I used `inspect()` to have more "public" access to it.

InstanceState.dict basically just return `__dict__`: https://github.com/sqlalchemy/sqlalchemy/blob/80576beab2918c1ac9e9e7e6fbc1e1a6cd5df237/lib/sqlalchemy/orm/base.py#L350

I guess it's fine either way :)

**Reply by zzzeek** [reply]
> 
> Do you suggest to switch to regular functions and not use any async functions? I'm wondering about performance then.. I would need to test that out.


it really depends on what the webapp is doing and what the requests are doing.   I think 99% of the time performance is not going to be different between asyncio and basic threaded, assuming the two approaches are essentially the same.  if the asyncio approach makes it easier to parallelize a bunch of IO/waiting work which would not be very straightforward in the non-asyncio approach, then the asyncio approach is helping with performance because it allowed you to organize the work better.  I use asyncio in some of my non-database work for this purpose and when it's the right fit it's terrific.

But I am getting ready to start doing serious benches of free-threaded python sometime in the foreseeable future, and I think if your CPU has more than one core (which is all of them) I think it should be easy to show that you can get way more throughput out of the python interpreter with real multi-core saturation, which is not possible in asyncio.    But ill wait to prove that theory before I spend too much time on that.

> 
> In the mean time, I came with such an idea for Pydantic transform problem, instead of passing DB model directly I use `inspect()` to get `dict` for the model and pass that to Pydantic schema class (implicitly when returning from function), `dict` is simpler and does not perform any unwanted attributes access outside AsyncSession 🤷‍♂️:

yeah, this is the "normal" way to do that sort of thing, to access another kind of object that represents the view you want in that case.  "normal" means, "how we'd do it in 1998"  (likely in C, Perl or Java though).   Which is not too popular in todays pydantic/everything-is-rust ecosystem

**Reply by CaselIT** [reply]
> I think it should be easy to show that you can get way more throughput out of the python interpreter with real multi-core saturation, which is not possible in asyncio.

I think I read somewhere that there were speedup also with asyncio in free-threaded python, but I can't seem to find the article and they may have been using `asyncio.to_thread()`
